remotes/mount.go

```diff
@@ -4,86 +4,80 @@ import (
 	"context"
 	"fmt"
 	"io"
-	"os"
+	"strings"
 
 	"github.com/containerd/containerd/content"
 	"github.com/containerd/containerd/errdefs"
 	"github.com/containerd/containerd/images"
 	"github.com/containerd/containerd/remotes"
-	"github.com/containerd/containerd/remotes/docker"
 	ocischemav1 "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
 )
 
 type imageHandler interface {
-	Handle(context.Context, ocischemav1.Descriptor) error
+	Handle(context.Context, *descriptorProgress) error
 }
 
-func newImageCopier(ctx context.Context, resolver docker.ResolverBlobMounter, sourceFetcher remotes.Fetcher, targetRepo string) (imageCopier, error) {
+func newImageCopier(ctx context.Context, resolver remotes.Resolver, sourceFetcher remotes.Fetcher, targetRepo string, eventNotifier eventNotifier) (imageCopier, error) {
 	destPusher, err := resolver.Pusher(ctx, targetRepo)
 	if err != nil {
 		return imageCopier{}, err
 	}
 	return imageCopier{
 		sourceFetcher: sourceFetcher,
 		targetPusher:  destPusher,
+		eventNotifier: eventNotifier,
 	}, nil
 }
 
 type imageCopier struct {
 	sourceFetcher remotes.Fetcher
 	targetPusher  remotes.Pusher
+	eventNotifier eventNotifier
 }
 
-func (h *imageCopier) Handle(ctx context.Context, desc ocischemav1.Descriptor) (err error) {
+func (h *imageCopier) Handle(ctx context.Context, desc *descriptorProgress) (retErr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	if len(desc.URLs) > 0 {
-		fmt.Fprintf(os.Stderr, "Skipping foreign descriptor %s with media type %s (size: %d)\n", desc.Digest, desc.MediaType, desc.Size)
+		desc.markDone()
+		desc.setAction("Skip (foreign layer)")
 		return nil
 	}
-	fmt.Fprintf(os.Stderr, "Copying descriptor %s with media type %s (size: %d)\n", desc.Digest, desc.MediaType, desc.Size)
-	reader, err := h.sourceFetcher.Fetch(ctx, desc)
-	if err != nil {
-		return err
-	}
-	defer reader.Close()
-	writer, err := h.targetPusher.Push(ctx, desc)
+	desc.setAction("Copy")
+	h.eventNotifier.reportProgress(nil)
+	defer func() {
+		h.eventNotifier.reportProgress(retErr)
+	}()
+	writer, err := h.targetPusher.Push(ctx, desc.Descriptor)
 	if err != nil {
 		if errors.Cause(err) == errdefs.ErrAlreadyExists {
+			desc.markDone()
+			if strings.Contains(err.Error(), "mounted") {
+				desc.setAction("Mounted")
+			}
 			return nil
 		}
+		desc.setError(err)
```

Comment:
Instead of calling `setError` each time, it would be good to factor this into the `defer`. The only case that you don't `setErr` when you return an error is when the error is `ErrAlreadyExists`.